arch/arm64/core/mmu.c

```diff
@@ -806,6 +806,79 @@ static uint64_t get_tcr(int el)
 	return tcr;
 }
 
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+
+static int flush_inv_mapped_pages(int (*cmo_fn)(void *, size_t),
+				  uint64_t *pte, int level, uint64_t base)
+{
+	int ret = 0;
+	uint64_t level_size = BIT(LEVEL_TO_VA_SIZE_SHIFT(level));
+	const uint64_t max_va_addr = (uint64_t)CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE;
+
+	if (!xlat_use_count[table_index(pte)]) {
+		return 0;
+	}
+
+	for (unsigned int i = 0; i < Ln_XLAT_NUM_ENTRIES; i++) {
+		uint64_t end, va = base + i * level_size;
+		uint8_t mem_type;
+
+		if (is_free_desc(pte[i])) {
+			continue;
+		}
+
+		if (is_table_desc(pte[i], level)) {
+			/* Move to the next translation table level */
+			ret = flush_inv_mapped_pages(cmo_fn, pte_desc_table(pte[i]), level + 1, va);
+			if (ret) {
+				break;
+			}
+			continue;
+		}
+
+		/*
+		 * From this point, this must be a leaf.
+		 * Start excluding non memory mappings
+		 */
+		mem_type = (pte[i] >> 2) & MT_TYPE_MASK;
+		if ((mem_type != MT_NORMAL) && (mem_type != MT_NORMAL_NC)) {
+			continue;
+		}
+
+		end = va + level_size - 1;
+
+		/* No intersection with RAM? */
+		if (end < CONFIG_KERNEL_VM_BASE || va >= max_va_addr) {
+			continue;
+		}
+
+		va = MAX(va, CONFIG_KERNEL_VM_BASE);
+		end = MIN(end, max_va_addr);
+
+		ret = cmo_fn(UINT_TO_POINTER(va), end - va);
+		if (ret) {
+			break;
+		}
+
+		MMU_DEBUG("Flush/inv PTE %llx at level %d: %llx-%llx\n",
+			  pte, level, va, end);
+	}
+
+	return ret;
+}
+
+int flush_inv_page_tbl_entries(int (*cmo_fn)(void *, size_t))
+{
+	uint64_t *base_xlat_table = UINT_TO_POINTER(read_ttbr0_el1());
+
+	if (!base_xlat_table) {
+		return -EINVAL;
+	}
+
+	return flush_inv_mapped_pages(cmo_fn, base_xlat_table, BASE_XLAT_LEVEL, 0);
```

Comment:
You are assuming that translation starts at level 0, which is incorrect. Depending on translation granule size and input address size, it can start at level 1 (and maybe even at level 2): paragraph FTBXR in ARM DDI 0487I.a

There is already `BASE_XLAT_LEVEL` defined in the Zephyr codebase

Reply:
Ooops, sorry. My bad.